File: SudokuPy/SudokuPy/solvers/DlxPuzzleSolver.py
from itertools import product
import math
from typing import Dict, List, Tuple, Union

from SudokuPy.Puzzle import Puzzle
from SudokuPy.solvers.PuzzleSolver import PuzzleSolver


class DlxPuzzleSolver(PuzzleSolver):
    """
    DLX support - inspired by Ali Assaf's Algorithm X in 30 lines! [1]
    [1]: URL https://www.cs.mcgill.ca/~aassaf9/python/algorithm_x.html
    """
    def __init__(self, p: Puzzle):
        """Initializer.

        Sets up the DLX constraints
        there are constraints that:
        - in each cell must be exactly one number
        - in each row must be all numbers
        - in each column must be all numbers
        - in each box must be all numbers

        Args:
            p (Puzzle): the Puzzle to be solved.
        """
        super().__init__(p)

        self.x = {c: set() for c in product(
            ['cell', 'row', 'column', 'box'],
            [v for v in product(range(1, self.puzzle.size + 1), range(1, self.puzzle.size + 1))])}

        box_size = int(math.sqrt(self.puzzle.size))
        self.y = {}
        for r, c, n in product(range(1, self.puzzle.size + 1),
                               range(1, self.puzzle.size + 1),
                               range(1, self.puzzle.size + 1)):
            b = ((r - 1) // box_size) * box_size + ((c - 1) // box_size) + 1
            self.y[(f'r{r}c{c}', f'{n}')] = [
                ("cell", (r, c)),
                ("row", (r, n)),
                ("column", (c, n)),
                ("box", (b, n))]

        for i, row in self.y.items():
            for j in row:
                self.x[j].add(i)

        # load grid
        for i in self.puzzle.grid:
            v = self.puzzle.grid[i]
            if v:
                self.cover(self.x, self.y, (i, v))

    def solve(self) -> Union[Puzzle, None]:
        """Main entry point for using DLX to solve a sudoku.

        Note that multiple solutions is an invalid puzzle!

        Returns:
            One or more solutions in a Puzzle
        """
        solution_count = 0
        for solution in self.recursive_solve(self.x, self.y, []):
            solution_count += 1
            if solution_count > 2:
                return
            # solution is a partial puzzle grid with only the name value pairs for the unknown cells

            # create a result that is a copy of the original puzzle
            result = Puzzle()
            for k, v in self.puzzle.grid.items():
                result.grid[k] = v

            # then add the solution to it
            for (i, n) in solution:
                result.grid[i] = n

            yield result

    def recursive_solve(self,
                        x: Dict[Tuple[str, Tuple[int, int]], set],
                        y: Dict[Tuple[str, str], List[Tuple[str, Tuple[int, int]]]],
                        solutions: List[Tuple[str, str]]) -> Union[List[Tuple[str, str]], None]:
        """ Implements the DLX solution algorithm

        Args:
            x: The DLX nodes
            y: The DLX row headers
            solutions: Buffer for solutions

        Returns:
            Solutions from deeper recursion
        """
        if not x:
            yield list(solutions)
        else:
            c = min(x, key=lambda i: len(x[i]))
            for r in list(x[c]):
                solutions.append(r)
                # put the covered columns on the stack
                cols = self.cover(x, y, r)
                for s in self.recursive_solve(x, y, solutions):
                    yield s
                self.uncover(x, y, r, cols)
                solutions.pop()

    @staticmethod
    def cover(x: Dict[Tuple[str, Tuple[int, int]], set],
              y: Dict[Tuple[str, str], List[Tuple[str, Tuple[int, int]]]],
              r: Tuple[str, str]) -> List[set]:
        """ DLX cover method - try a solution

        Args:
            x: The DLX nodes
            y: The DLX row headers
            r: The DLX row to eliminate

        Returns:
            The DLX nodes from the column eliminated
        """
        cols = []
        for j in y[r]:
            for i in x[j]:
                for k in y[i]:
                    if k != j:
                        x[k].remove(i)
            cols.append(x.pop(j))
        return cols

    @staticmethod
    def uncover(x: Dict[Tuple[str, Tuple[int, int]], set],
                y: Dict[Tuple[str, str], List[Tuple[str, Tuple[int, int]]]],
                r: Tuple[str, str],
                cols: List[set]):
        """ DLX uncover method - restores a row header and column nodes to the model.

        Args:
            x: The DLX nodes
            y: The DLX row headers
            r: The DLX row to restore
            cols: The DLX nodes from the column restored

        Returns:
            None
        """
        for j in reversed(y[r]):
            x[j] = cols.pop()
            for i in x[j]:
                for k in y[i]:
                    if k != j:
                        x[k].add(i)

File: SudokuTs/src/solvers/DlxPuzzleSolver.ts
import * as Collections from 'typescript-collections';
import { Puzzle } from '../Puzzle';
import { PuzzleSize, PuzzleSizeUtil } from '../PuzzleSize';
import { product2, product3, range } from './IterationUtils';

export class DlxPuzzleSolver {
    public puzzle: Puzzle;

    private x: Collections.Dictionary<[string, [number, number]], Set<[string, string]>>
        = new Collections.Dictionary<[string, [number, number]], Set<[string, string]>>();
    private y: Collections.Dictionary<[string, string], Array<[string, [number, number]]>>
        = new Collections.Dictionary<[string, string], Array<[string, [number, number]]>>();

    constructor(puzzle: Puzzle) {
        this.puzzle = puzzle;

        const constraintTypes: string[] = ['cell', 'row', 'column', 'box'];
        const valueRange: number[] = range(1, this.puzzle.puzzleSize + 1);
        constraintTypes.forEach(constraint => {
            product2(valueRange, valueRange).forEach(pair => {
                this.x.setValue([constraint, pair], new Set<[string, string]>());
            });
        });

        const boxSize: number = Math.sqrt(this.puzzle.puzzleSize);
        product3(valueRange, valueRange, valueRange).forEach(triple => {
            const [r, c, n] = triple;

            const yKey: [string, string] = [`r${r}c${c}`, `${n}`];
            const b = Math.floor((r - 1) / boxSize) * boxSize + Math.floor((c - 1) / boxSize) + 1;

            const yValue: Array<[string, [number, number]]>
                = [['cell', [r, c]], ['row', [r, n]], ['column', [c, n]], ['box', [b, n]]];

            this.y.setValue(yKey, yValue);
        });

        this.y.keys().forEach(i => {
            const row = this.y.getValue(i);
            row?.forEach(j => {
                this.x.getValue(j)?.add(i);
            });
        });

        // load the puzzle
        this.puzzle.grid.keys().forEach(i => {
            const v = this.puzzle.grid.getValue(i);
            if (v !== null && v !== undefined) {
                this.cover(this.x, this.y, [i, v]);
            }
        });
    }

    public solve(): Puzzle[] {
        const solutions = new Array<Puzzle>();

        for (const solution of this.recursiveSolve(this.x, this.y, new Array<[string, string]>())) {
                if (solutions.length >= 2) {
                    return solutions;
                }
                // solution is a partial puzzle grid with only the name value pairs for the unknown cells

                // create a result that is a copy of the original puzzle
                const result = new Puzzle(this.puzzle);

                // then add the solution to it
                for (const s in solution) {
                    if (s !== undefined) {
                        const [i, n] = s;
                        result.grid.setValue(i, n);
                    }
                }

                solutions.push(result);
            }

        return solutions;
    }

    private *recursiveSolve(
        x: Collections.Dictionary<[string, [number, number]], Set<[string, string]>>,
        y: Collections.Dictionary<[string, string], Array<[string, [number, number]]>>,
        solution: Array<[string, string]>): Generator<Array<[string, string]>> {

        if (x.isEmpty()) {
            yield solution;
        } else {
            // find smallest set in x

            let minSetSize = this.puzzle?.puzzleSize ?? 9;
            // this finds smallest set size
            // for (const p in x.keys()) {
            //     if (x.getValue(p) !== undefined) {
            //         const thisSize = x.getValue(p)?.size ?? 0;
            //         if (thisSize < minSetSize) {
            //             minSetSize = thisSize;
            //         }
            //     }
            // }

            x.keys().forEach( p => {
                const nextSize = x.getValue(p)?.size ?? 9;
                if (nextSize < minSetSize) {
                    minSetSize = nextSize;
                }
            });

            let c: [string, [number, number]] | undefined;

            x.keys().some(i => {
                if (x.getValue(i)?.size === minSetSize) {
                    c = i;
                    return true;
                }
            });

            if (c !== undefined) {
                const xSets: Array<[string, string]>|undefined = [...x.getValue(c)];
                if (xSets !== undefined) {
                    // tslint:disable-next-line: prefer-for-of
                    for (let i = 0; i < xSets.length; i++) {
                        const r = xSets[i];
                        solution.push(r);
                        // put the covered columns on the stack
                        const cols = this.cover(x, y, r);
                        // tslint:disable-next-line: forin
                        const values = [...this.recursiveSolve(x, y, solution)];
                        // tslint:disable-next-line: prefer-for-of
                        for (let j = 0; j < values.length; j++) {
                            yield values[j];
                        }
                        this.uncover(x, y, r, cols);
                        solution.pop();
                    }
                }
            }
        }
    }

    private cover(
        x: Collections.Dictionary<[string, [number, number]], Set<[string, string]>>,
        y: Collections.Dictionary<[string, string], Array<[string, [number, number]]>>,
        r: [string, string]): Array<Set<[string, string]>> {

        const cols: Array<Set<[string, string]>> = [];
        y.getValue(r)?.forEach(j => {
            x.getValue(j)?.forEach(i => {
                y.getValue(i)?.forEach(k => {
                    if (JSON.stringify(k) !== JSON.stringify(j)) {
                        x.getValue(k)?.delete(i);
                    }

                    const v = x.getValue(j);
                    x.remove(j);
                    if (v !== undefined) {
                        cols.push(v);
                    }
                });
            });
        });

        return cols;
    }

    private uncover(
        x: Collections.Dictionary<[string, [number, number]], Set<[string, string]>>,
        y: Collections.Dictionary<[string, string], Array<[string, [number, number]]>>,
        r: [string, string],
        cols: Array<Set<[string, string]>>) {

        y.getValue(r)?.slice().reverse().forEach(j => {
            const v = cols.pop();
            if (v !== undefined) {
                x.setValue(j, v);
            }

            x.getValue(j)?.forEach(i => {
                y.getValue(i)?.forEach(k => {
                    if (JSON.stringify(k) !== JSON.stringify(j)) {
                        x.getValue(k)?.add(i);
                    }
                });
            });
        });

        return cols;
    }
}
